avahi-core/server.c

```diff
@@ -1626,30 +1638,42 @@ AvahiServerConfig* avahi_server_config_copy(AvahiServerConfig *ret, const AvahiS
             return NULL;
 
     if (c->domain_name)
-        if (!(d = avahi_strdup(c->domain_name))) {
-            avahi_free(h);
-            return NULL;
+        if (!(d = avahi_strdup(c->domain_name)))
+            goto clean;
+
+    if (!(browse = avahi_string_list_copy(c->browse_domains)) && c->browse_domains)
+        goto clean;
+
+    if (!(allow = avahi_string_list_copy(c->allow_interfaces)) && c->allow_interfaces)
+        goto clean;
+
+    if (!(deny = avahi_string_list_copy(c->deny_interfaces)) && c->deny_interfaces)
+        goto clean;
+
+    if (c->allow_addresses) {
+        if (!(allow_addresses = avahi_new(AvahiAddressList, 1)))
+            goto clean;
+        AVAHI_LLIST_INIT(AvahiAddressList, address, allow_addresses);
+        for (l = c->allow_addresses; l; l = l->address_next) {
+            if (!(next = avahi_new(AvahiAddressList, 1)))
+                goto clean;
+            AVAHI_LLIST_INIT(AvahiAddressList, address, next);
+            next->address = l->address;
+            AVAHI_LLIST_PREPEND(AvahiAddressList, address, allow_addresses, next);
         }
-
-    if (!(browse = avahi_string_list_copy(c->browse_domains)) && c->browse_domains) {
-        avahi_free(h);
-        avahi_free(d);
-        return NULL;
     }
 
-    if (!(allow = avahi_string_list_copy(c->allow_interfaces)) && c->allow_interfaces) {
-        avahi_string_list_free(browse);
-        avahi_free(h);
-        avahi_free(d);
-        return NULL;
-    }
-
-    if (!(deny = avahi_string_list_copy(c->deny_interfaces)) && c->deny_interfaces) {
-        avahi_string_list_free(allow);
-        avahi_string_list_free(browse);
-        avahi_free(h);
-        avahi_free(d);
-        return NULL;
+    if (c->deny_addresses) {
+        if (!(deny_addresses = avahi_new(AvahiAddressList, 1)))
+            goto clean;
+        AVAHI_LLIST_INIT(AvahiAddressList, address, allow_addresses);
```

Comment:
Incorrectly passing 'allow_addresses' into the list init instead of 'deny_addresses'.

Reply:
Yes, thank you for discovering this. This seems classic copy&paste error, which suggests we want to have ``avahi_address_list_copy`` function, where just allow_addresses and deny_addresses are passed as parameter to initialize.